lib/BloqadeQMC/src/updates/cluster.jl

```diff
@@ -0,0 +1,333 @@
+            # return logweightchange(qmc_state.trialstate, qmc_state.left_config[i])
+            return 0.0
+        elseif i > (lsize - Ns)
+            # return logweightchange(qmc_state.trialstate, qmc_state.right_config[i])
+            return 0.0
+        else
+            return 0.0
+        end
+    end
+end
+trialstate_weight_change(qmc_state::BinaryThermalState, lsize::Int, Ns::Int, i::Int) = 0.0
+
+#############################################################################
+
+# changed the function name below from cluster_update!() to generic_cluster_update!()
+
+function generic_cluster_update!(rng::AbstractRNG, update_kernel!::Function, acceptance::Function, lsize::Int, qmc_state::BinaryQMCState, H::AbstractIsing, d::Diagnostics)
+    Ns = nspins(H)
+    operator_list = qmc_state.operator_list
+
+    LinkList = qmc_state.linked_list
+    LegType = qmc_state.leg_types
+    Associates = qmc_state.associates
+    op_indices = qmc_state.op_indices
+
+    in_cluster = fill!(qmc_state.in_cluster, 0)
+    cstack = qmc_state.cstack # This is the stack of vertices in a cluster
+    current_cluster = qmc_state.current_cluster
+    runstats = d.runstats
+
+    ccount = 0  # cluster number counter
+
+    num_accept = 0
+    num_reject = 0
+
+    @inbounds for i in 1:lsize
+        # Add a new leg onto the cluster
+        if iszero(in_cluster[i]) && iszero(Associates[i])
+            ccount += 1
+            push!(cstack, i)
+            in_cluster[i] = ccount
+            if qmc_state isa BinaryGroundState && !(qmc_state.trialstate isa AbstractProductState)
+                left_flips = empty!(qmc_state.trialstate.left_flips)
+                right_flips = empty!(qmc_state.trialstate.right_flips)
+            end
+
+            empty!(current_cluster)
+            push!(current_cluster, i)
+            lnA = 0.0
+            if qmc_state isa BinaryGroundState
+                lnA += trialstate_weight_change(qmc_state, lsize, Ns, i)
+            end
+
+            while !isempty(cstack)
+                leg = LinkList[pop!(cstack)]
+
+                if iszero(in_cluster[leg])
+                    in_cluster[leg] = ccount  # add the new leg and flip it
+                    push!(current_cluster, leg)
+                    if qmc_state isa BinaryGroundState
+                        lnA += trialstate_weight_change(qmc_state, lsize, Ns, i)
+                    end
+                    a = Associates[leg]
+
+                    a == 0 && continue
+                    # from this point on, we know we're on a bond op
+                    op = operator_list[op_indices[leg]]
+                    w = getweightindex(H, op) - getoperatortype(H, op)
+                    preflip_bond_type, postflip_bond_type = update_kernel!(qmc_state, H, ccount, leg, a)
+                    lnA += (
+                        getlogweight(H.op_sampler, w + postflip_bond_type)
+                        - getlogweight(H.op_sampler, w + preflip_bond_type)
+                    )
```

Comment:
I'd separate this to a function (at least to start with) it's always beneficial to wrap a loop kernel into a function